src/jvmMain/kotlin/io/beatmaps/api/bookmark.kt

```diff
@@ -0,0 +1,157 @@
+package io.beatmaps.api
+
+import de.nielsfalk.ktor.swagger.DefaultValue
+import de.nielsfalk.ktor.swagger.Ignore
+import de.nielsfalk.ktor.swagger.get
+import de.nielsfalk.ktor.swagger.ok
+import de.nielsfalk.ktor.swagger.post
+import de.nielsfalk.ktor.swagger.responds
+import de.nielsfalk.ktor.swagger.version.shared.Group
+import io.beatmaps.common.api.EPlaylistType
+import io.beatmaps.common.dbo.Beatmap
+import io.beatmaps.common.dbo.Playlist
+import io.beatmaps.common.dbo.PlaylistMap
+import io.beatmaps.common.dbo.User
+import io.beatmaps.common.dbo.complexToBeatmap
+import io.beatmaps.util.cdnPrefix
+import io.ktor.http.HttpStatusCode
+import io.ktor.server.application.call
+import io.ktor.server.locations.Location
+import io.ktor.server.locations.options
+import io.ktor.server.response.header
+import io.ktor.server.response.respond
+import io.ktor.server.routing.Route
+import org.jetbrains.exposed.sql.and
+import org.jetbrains.exposed.sql.deleteWhere
+import org.jetbrains.exposed.sql.insert
+import org.jetbrains.exposed.sql.insertAndGetId
+import org.jetbrains.exposed.sql.select
+import org.jetbrains.exposed.sql.transactions.transaction
+import org.jetbrains.exposed.sql.update
+
+@Location("/api")
+class BookmarksApi {
+    @Group("Bookmarks") @Location("/bookmarks/add")
+    data class Add(@Ignore val api: BookmarksApi)
+    @Group("Bookmarks") @Location("/bookmarks/remove")
+    data class Remove(@Ignore val api: BookmarksApi)
+    @Group("Bookmarks") @Location("/bookmarks/{page}")
+    data class Bookmarks(@Ignore val api: BookmarksApi, @DefaultValue("0") val page: Long = 0)
+}
+
+fun getBookmarksId(userId: Int): Int {
+    var id = User
+        .slice(User.bookmarksId)
+        .select { User.id eq userId }
+        .first()[User.bookmarksId]?.value
+
+    if (id == null) {
+        id = Playlist.insertAndGetId {
+            it[type] = EPlaylistType.System
+            it[name] = "Bookmarks"
+            it[description] = "Maps you've bookmarked, automatically updated."
+            it[owner] = userId
+        }.value
+
+        User.update({ User.id eq userId }) {
+            it[bookmarksId] = id
+        }
+    }
+
+    return id
+}
+
+fun isBookMarked(mapId: Int, userId: Int): Boolean {
+    val bookmarksId = getBookmarksId(userId)
+
+    return PlaylistMap
+        .select { (PlaylistMap.mapId eq mapId) and (PlaylistMap.playlistId eq bookmarksId) }
+        .count() > 0
+}
+
+fun Route.bookmarkRoute() {
+    options<BookmarksApi.Add> {
+        call.response.header("Access-Control-Allow-Origin", "*")
```

Comment:
What's the rationale for having cors headers for these endpoints?
afaik we still don't have an auth strategy for mods (which wouldn't need these) and the site itself doesn't need them.

Including them in the swagger docs is also questionable as they're authenticated endpoints (only reason is for us to point mod authors at, but we can't do that yet as no auth)

Reply:
I thought these endpoints would need to be public so they could be used by BeatSync.

Reply:
CORS headers only apply to browsers, beatsync doesn't need them.

Reply:
Does the same go for the option endpoints?

Reply:
Yes